x/superfluid/keeper/stake.go

```diff
@@ -350,3 +350,64 @@ func (k Keeper) forceUndelegateAndBurnOsmoTokens(ctx sdk.Context,
 // Eugen’s point: Only rewards message needs to be updated. Rest of messages are fine
 // Queries need to be updated
 // We can do this at the very end though, since it just relates to queries.
+
+// IterateBondedValidatorsByPower implements govtypes.StakingKeeper
+func (k Keeper) IterateBondedValidatorsByPower(ctx sdk.Context, fn func(int64, stakingtypes.ValidatorI) bool) {
+	k.sk.IterateBondedValidatorsByPower(ctx, fn)
+}
+
+// TotalBondedTokens implements govtypes.StakingKeeper
+func (k Keeper) TotalBondedTokens(ctx sdk.Context) sdk.Int {
+	return k.sk.TotalBondedTokens(ctx)
+}
+
+// IterateDelegations implements govtypes.StakingKeeper
+func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn func(int64, stakingtypes.DelegationI) bool) {
+	synthlocks := k.lk.GetAllSyntheticLockupsByAddr(ctx, delegator)
+	for i, lock := range synthlocks {
+		// get locked coin from the lock ID
+		interim, ok := k.GetIntermediaryAccountFromLockId(ctx, lock.UnderlyingLockId)
+		if !ok {
+			panic(fmt.Sprintf("intermediary account retrieval failed with underlying lock(Lock: %+v)", lock))
+		}
+		lock, err := k.lk.GetLockByID(ctx, lock.UnderlyingLockId)
+		if err != nil {
+			panic(fmt.Sprintf("lockup retrieval failed with underlying lock(Lock: %+v; Error: %s)", lock, err))
+		}
+		coin, err := lock.SingleCoin()
+		if err != nil {
+			panic(fmt.Sprintf("no single coin in the lock(Lock: %+v; Error: %s)", lock, err))
+		}
+
+		// get osmo-equivalent token amount
+		amount := k.GetSuperfluidOSMOTokens(ctx, interim.Denom, coin.Amount)
+
+		// get validator shares equivalent to the token amount
+		valAddr, err := sdk.ValAddressFromBech32(interim.ValAddr)
+		if err != nil {
+			panic(fmt.Sprintf("validator address decoding failed(Lock: %+v; Address: %s; Error: %s)", lock, interim.ValAddr, err))
+		}
+		validator, found := k.sk.GetValidator(ctx, valAddr)
+		if !found {
+			panic(fmt.Sprintf("validator not exists(Lock: %+v; Address: %s)", lock, valAddr))
+		}
+		shares, err := validator.SharesFromTokens(amount)
+		if err != nil {
+			// tokens are not valid. continue.
+			continue
+		}
+
+		// construct delegation and call callback
+		delegation := stakingtypes.Delegation{
+			DelegatorAddress: delegator.String(),
+			ValidatorAddress: interim.ValAddr,
+			Shares:           shares,
+		}
+		fn(int64(i), delegation)
+	}
+
+	// call the callback with the non-superfluid delegations
+	k.sk.IterateDelegations(ctx, delegator, func(i int64, delegation stakingtypes.DelegationI) (stop bool) {
+		return fn(int64(len(synthlocks))+i, delegation) // add len(synthlocks) to index to avoid overlapping
```

Comment:
does using lengths of synth locks ensure that we avoid overlapping? (does synth delegation come first?)

Reply:
I put the synth delegation first because the index itself was just for incrementing counter, but it feels like preserving the original index is more safe.